mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageMapper.java

```diff
@@ -0,0 +1,373 @@
+                .then(Mono.just(map)));
+    }
+
+    @Override
+    public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) {
+        return mailboxMessageDAO.findFirstUnseenMessageUid((PostgresMailboxId) mailbox.getMailboxId()).block();
+    }
+
+    @Override
+    public Mono<Optional<MessageUid>> findFirstUnseenMessageUidReactive(Mailbox mailbox) {
+        return mailboxMessageDAO.findFirstUnseenMessageUid((PostgresMailboxId) mailbox.getMailboxId())
+            .map(Optional::of)
+            .switchIfEmpty(Mono.just(Optional.empty()));
+    }
+
+    @Override
+    public List<MessageUid> findRecentMessageUidsInMailbox(Mailbox mailbox) {
+        return findRecentMessageUidsInMailboxReactive(mailbox).block();
+    }
+
+    @Override
+    public Mono<List<MessageUid>> findRecentMessageUidsInMailboxReactive(Mailbox mailbox) {
+        return mailboxMessageDAO.findAllRecentMessageUid((PostgresMailboxId) mailbox.getMailboxId())
+            .collectList();
+    }
+
+    @Override
+    public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        return addReactive(mailbox, message).block();
+    }
+
+    @Override
+    public Mono<MessageMetaData> addReactive(Mailbox mailbox, MailboxMessage message) {
+        return Mono.fromCallable(() -> {
+                message.setSaveDate(Date.from(clock.instant()));
+                return message;
+            })
+            .flatMap(mailboxMessage -> setNewMessageUid(mailboxMessage)
+                .then(setNewMessageModSeq(mailboxMessage)))
+            .then(mailboxMessageDAO.insert(message))
+            .then(saveBodyContent(message)
+                .flatMap(blobId -> messageDAO.insert(message, blobId.asString())))
+            .thenReturn(message.metaData());
+    }
+
+    private Mono<BlobId> saveBodyContent(MailboxMessage message) {
+        return Mono.fromCallable(() -> MESSAGE_CONTENT_LOADER.apply(message))
+            .flatMap(bodyByteSource -> Mono.from(blobStore.save(blobStore.getDefaultBucketName(), bodyByteSource, LOW_COST)));
+    }
+
+    @Override
+    public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator, MessageRange range) {
+        return updateFlagsPublisher(mailbox, flagsUpdateCalculator, range)
+            .toIterable()
+            .iterator();
+    }
+
+    @Override
+    public Mono<List<UpdatedFlags>> updateFlagsReactive(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator, MessageRange range) {
+        return updateFlagsPublisher(mailbox, flagsUpdateCalculator, range)
+            .collectList();
+    }
+
+    private Flux<UpdatedFlags> updateFlagsPublisher(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator, MessageRange range) {
+        return mailboxMessageDAO.findMessagesMetadata((PostgresMailboxId) mailbox.getMailboxId(), range)
+            .flatMap(currentMetaData -> updateFlags(currentMetaData, flagsUpdateCalculator));
+    }
```

Comment:
> One modseq for all

I did it in last commit 

> Please let's update all flags at once here too

We can  not update flags in "single" query for all messages, because the `newFlags` of each message is a difference

Reply:
> We can not update flags in "single" query for all messages, because the newFlags of each message is a difference

I know, sorry for being unclear.